pkg/dns/service.go

```diff
@@ -224,28 +224,38 @@ func (s *Service) SetEndpoints(ctx context.Context, addresses []gatewayv1beta1.G
 
 // GetManagedZoneForHost returns a ManagedZone and subDomain for the given host if one exists.
 //
-// Currently, this returns the first matching ManagedZone found in the traffic resources own namespace, or if none is found,
-// it looks for the first matching one listed in the default ctrl namespace.
+// Currently, this returns the first matching ManagedZone found in the traffic resources own namespace
 func (s *Service) GetManagedZoneForHost(ctx context.Context, host string, t traffic.Interface) (*v1alpha1.ManagedZone, string, error) {
-	hostParts := strings.SplitN(host, ".", 2)
-	if len(hostParts) < 2 {
-		return nil, "", fmt.Errorf("unable to parse host : %s on traffic resource : %s", host, t.GetName())
-	}
-	subDomain := hostParts[0]
-	parentDomain := hostParts[1]
-
 	var managedZones v1alpha1.ManagedZoneList
-
-	if err := s.controlClient.List(ctx, &managedZones, client.InNamespace(t.GetNamespace()), client.MatchingFields{"spec.domainName": parentDomain}); err != nil {
+	if err := s.controlClient.List(ctx, &managedZones, client.InNamespace(t.GetNamespace())); err != nil {
 		log.FromContext(ctx).Error(err, "unable to list managed zones in traffic resource NS")
 		return nil, "", err
 	}
+	return FindMatchingManagedZone(host, host, managedZones.Items)
+}
+
+func FindMatchingManagedZone(originalHost, host string, zones []v1alpha1.ManagedZone) (*v1alpha1.ManagedZone, string, error) {
+	if len(zones) == 0 {
+		return nil, "", fmt.Errorf("no managed zone found for host : %s", host)
+	}
+	host = strings.ToLower(host)
+	hostParts := strings.SplitN(host, ".", 2)
```

Comment:
Do we need to consider tldrs that are more than 1 i.e. co.uk. It maybe doesn't matter here really, but we should probably prevent a MZ from being created with a domain that isn't at least a root domain i.e <domain name>.<top level domain>

Reply:
good catch, fixed this now.